llvm/test/CodeGen/SPIRV/pointers/global-addrspacecast.ll

```diff
@@ -0,0 +1,17 @@
+; RUN: llc -verify-machineinstrs -O0 -mtriple=spirv64-unknown-unknown %s -o - | FileCheck %s
+; RUN: %if spirv-tools %{ llc -O0 -mtriple=spirv64-unknown-unknown %s -o - -filetype=obj | spirv-val %}
+
+@PrivInternal = internal addrspace(10) global i32 456
+; CHECK-DAG:  %[[#type:]] = OpTypeInt 32 0
+; CHECK-DAG: %[[#ptrty:]] = OpTypePointer Private %[[#type]]
+; CHECK-DAG: %[[#value:]] = OpConstant %[[#type]] 456
+; CHECK-DAG:   %[[#var:]] = OpVariable %[[#ptrty]] Private %[[#value]]
+
+define spir_kernel void @Foo() {
+  %p = addrspacecast ptr addrspace(10) @PrivInternal to ptr
```

Comment:
This test is specifically to validate we can lower an addrspacecast from private to function.

This pattern can be generated by the actual codegen:
- https://clang.llvm.org/doxygen/CodeGenModule_8cpp_source.html#l05259
- https://clang.llvm.org/doxygen/CodeGenModule_8cpp_source.html#l04982

What happens is a global is recreated with the correct AS, and where a load was done (or function call with this as parameter), the `addrspacecast` is added.
In basic cases, I don't think this is an issue, as it generates a `load(addrspacecast(ptr))`.

BUT I suspect if we start to have references, then it could become an issue, as the ptr will be changed to another AS, and suddenly the `OpTypePtr` becomes `Function` instead of `Private`?

Reply:
Okay, we can leave this for now. I'll want to see how this code pattern is generated by clang/llvm because I fear it could lead to invalid spir-v, but we can deal with that when we start using the address space in clang.

Reply:
Yes, I actually have 1 pattern that creates a bad pattern before inline. Trying to find a good way, but haven't so far.
But that's for the FE so fine for this PR.